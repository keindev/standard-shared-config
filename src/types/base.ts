export enum FileType {
  Text = 'text',
  GLOB = 'glob',
  JSON = 'json',
  YAML = 'yaml',
}

export enum EntityName {
  Scripts = 'scripts',
  Snapshots = 'snapshots',
  Package = 'package',
}

export type IMergeRule = string[] | null | boolean;
export type IScript = [string, string];
export type IMergeMap = { [key: string]: IMergeRule };

export interface ISnapshot {
  content: string;
  executable: boolean;
  hash: string;
  merge: IMergeRule;
  path: string;
  type: FileType;
}
