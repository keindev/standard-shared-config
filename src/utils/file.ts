import crypto from 'crypto';
import { constants, promises as fs } from 'fs';
import glob from 'glob';
import path from 'path';

import { FileType, IMergeRule, ISharedConfig, ISnapshot } from '../types.js';
import { merge, parse, stringify } from './json.js';

enum FileMode {
  Default = 0o0666,
  Executable = 0o0755,
}

const GLOB_OPTIONS = { dot: true, strict: true, nodir: true, ignore: ['**/.sharedconfig.yml'] };

export const readFile = async (filePath: string): Promise<string | undefined> => {
  const isExists = await fs
    .access(filePath, constants.R_OK)
    .then(() => true)
    .catch(() => false);
  let content: string | undefined;

  if (isExists) content = await fs.readFile(filePath, 'utf8');

  return content;
};

export const getHash = (data: string): string => {
  const hash = crypto.createHash('sha256');

  return hash.update(data, 'utf8').digest('hex');
};

export const getType = (filePath: string): FileType => {
  const extname = path.extname(filePath);
  const normalizedExtname = extname.substring(1).toLowerCase();
  let type;

  switch (true) {
    case normalizedExtname === 'json':
      type = FileType.JSON;
      break;
    case ['yaml', 'yml'].includes(normalizedExtname):
      type = FileType.YAML;
      break;
    case /\.[a-z]+ignore/i.test(path.basename(filePath, extname)):
      type = FileType.GLOB;
      break;
    default:
      type = FileType.Text;
      break;
  }

  return type;
};

export const writeFile = async (filePath: string, data: string | string[], executable = false): Promise<void> => {
  await fs.mkdir(path.dirname(filePath), { recursive: true });
  await fs.writeFile(filePath, Array.isArray(data) ? data.join('\n') : data, {
    mode: executable ? FileMode.Executable : FileMode.Default,
  });
};

export const createSnapshot = async (
  filePath: string,
  sharedDir: string,
  mergeRules: Map<string, IMergeRule>,
  executableFiles: Set<string>
): Promise<ISnapshot> => {
  const content = (await readFile(filePath)) ?? '';
  const relativePath = path.relative(sharedDir, filePath);
  const type = getType(filePath);
  const rule = mergeRules.get(relativePath);

  return {
    path: relativePath,
    hash: getHash(content),
    merge: [FileType.GLOB, FileType.Text].includes(type) && rule ? true : rule ?? false,
    executable: executableFiles.has(relativePath),
    type,
    content,
  };
};

export const createSnapshots = async (
  dir: string,
  config: Pick<ISharedConfig, 'include' | 'mergeRules' | 'executableFiles' | 'ignorePatterns'>
): Promise<ISnapshot[]> => {
  const files = (config.include ?? ['**/*']).map(pattern => glob.sync(`${dir}/${pattern}`, GLOB_OPTIONS)).flat();
  const mergeRules = new Map(
    Object.entries(config.mergeRules ?? {}).reduce((acc, [key, value]) => {
      if (Array.isArray(value)) acc.push([key, value.filter(item => typeof item === 'string')]);
      if (value === null) acc.push([key, !value]);
      if (typeof value === 'boolean') acc.push([key, value]);

      return acc;
    }, [] as [string, IMergeRule][])
  );
  const executableFiles = new Set(config.executableFiles ?? []);
  const snapshots = await Promise.all(
    files
      .sort((a, b) => b.split('/').length - a.split('/').length || a.localeCompare(b))
      .map(filePath => createSnapshot(filePath, dir, mergeRules, executableFiles))
  );

  return [
    ...snapshots,
    ...Object.entries(config.ignorePatterns ?? {}).map(([filePath, rows]) => {
      const content = rows.join('\n');

      return {
        path: filePath,
        hash: getHash(content),
        merge: !!mergeRules.get(filePath),
        executable: executableFiles.has(filePath),
        type: FileType.GLOB,
        content,
      };
    }),
  ];
};

export const mergeFiles = ({ type, merge: rules, content: snapshot }: ISnapshot, content: string): string => {
  const excludes = Array.isArray(rules) ? new Set(rules) : undefined;
  let result;

  switch (type) {
    case FileType.GLOB:
      result = [...new Set([...content.split('\n'), ...snapshot.split('\n')].filter(Boolean)).values()]
        .sort()
        .join('\n');
      break;
    case FileType.JSON:
    case FileType.YAML:
      result = stringify(merge({ left: parse(content, type), right: parse(snapshot, type), excludes }), type);
      break;
    case FileType.Text:
    default:
      result = content;
      break;
  }

  return result;
};
