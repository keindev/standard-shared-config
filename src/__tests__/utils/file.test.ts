// @see https://github.com/facebook/jest/issues/9430
import { promises as fs } from 'fs';
import { glob } from 'glob';

// eslint-disable-next-line node/no-extraneous-import
import { jest } from '@jest/globals';

import { FileType, IMergeRule } from '../../types/base.js';
import { createSnapshot, createSnapshots, getHash, getType, mergeFiles, readFile } from '../../utils/file.js';

enum FilePath {
  BASE = 'file.yml',
  SNAPSHOT = 'snapshot.json',
}

const DIR = '.config';
const CONTENT = {
  [FilePath.BASE]: `
  // Comment 1
  # Comment 2
  /* Comment 3 */
  test: true
  `,
  [FilePath.SNAPSHOT]: JSON.stringify({
    configurations: [
      {
        name: 'Build configs map',
        type: 'node',
        request: 'launch',
        outputCapture: 'std',
        console: 'integratedTerminal',
        runtimeExecutable: 'node',
        runtimeArgs: ['--nolazy', '--inspect-brk=9229', '--loader', 'ts-node/esm', 'src/cli/bin.ts', 'build'],
      },
    ],
  }),
};

jest.spyOn(fs, 'access').mockImplementation(() => Promise.resolve());
jest.spyOn(fs, 'chmod').mockImplementation(() => Promise.resolve());
jest.spyOn(fs, 'readFile').mockImplementation(path => Promise.resolve(CONTENT[path as FilePath] ?? ''));
jest.spyOn(glob, 'sync').mockImplementation(() => [FilePath.BASE, FilePath.SNAPSHOT]);

describe('File utils', () => {
  it('Get file content', async () => {
    const content = await readFile(FilePath.BASE);

    expect(content).toBe(CONTENT[FilePath.BASE]);
  });

  it('Get file hash', () => {
    expect(getHash(CONTENT[FilePath.BASE])).toMatchSnapshot();
  });

  it('Get file type', () => {
    expect(getType('.gitignore')).toBe(FileType.GLOB);
    expect(getType('test.json')).toBe(FileType.JSON);
    expect(getType('test')).toBe(FileType.Text);
    expect(getType('test.yaml')).toBe(FileType.YAML);
    expect(getType('test.yml')).toBe(FileType.YAML);
  });

  it('Create snapshot', async () => {
    const relativePath = `../${FilePath.SNAPSHOT}`;
    const mergeRules = new Map<string, IMergeRule>([[relativePath, ['configurations']]]);
    const snapshotBase = {
      path: relativePath,
      hash: 'd3ad7d184db5cd15a795d0713fe5c5115743f52d44776a130ed53a5e3a60124d',
      type: FileType.JSON,
      content: CONTENT[FilePath.SNAPSHOT],
    };

    const snapshot1 = await createSnapshot(FilePath.SNAPSHOT, DIR, new Map(), new Set());
    const snapshot2 = await createSnapshot(FilePath.SNAPSHOT, '.config', mergeRules, new Set([relativePath]));

    expect(snapshot1).toEqual({ ...snapshotBase, merge: false, executable: false });
    expect(snapshot2).toEqual({ ...snapshotBase, merge: mergeRules.get(relativePath), executable: true });
  });

  it('Create snapshots', async () => {
    const relativePath = `../${FilePath.SNAPSHOT}`;
    const mergeRules = ['configurations'];
    const snapshots = await createSnapshots(DIR, { include: ['.shared'], mergeRules: { [relativePath]: mergeRules } });

    expect(snapshots).toMatchSnapshot();
  });

  it('Merge JSON files', () => {
    const source = {
      text: 'text',
      object1: {
        value1: false,
        value2: true,
      },
      object2: {
        value1: true,
        value2: false,
        value3: null,
      },
      list: [1, 2, 3],
    };
    const current = {
      ...source,
      object1: {
        value1: true,
        value2: false,
        value3: null,
      },
      list: [1, 2, 3, 4, 5],
    };

    expect(
      mergeFiles(
        {
          path: '',
          hash: '',
          executable: false,
          type: FileType.JSON,
          merge: ['object1.value1', 'list'],
          content: JSON.stringify(source),
        },
        JSON.stringify(current)
      )
    ).toMatchSnapshot();
  });

  it('Merge YAML files', () => {
    const source = `
    version: 2
    updates:
    - package-ecosystem: npm
      directory: "/"
      schedule:
        interval: weekly
        day: friday
        time: "22:00"
    `;
    const current = `
    version: 2
    updates:
    - package-ecosystem: yarn
      schedule:
        interval: weekly
        time: "23:00"
    `;

    expect(
      mergeFiles(
        {
          path: '',
          hash: '',
          executable: false,
          type: FileType.YAML,
          merge: ['updates.0.schedule.time', 'updates.0.package-ecosystem'],
          content: source,
        },
        current
      )
    ).toMatchSnapshot();
  });

  it('Merge GLOB files', () => {
    const source = `
      *.log
      *.pid
      *.pid.lock
      *.seed
      *.tgz
      .editorconfig
      .env
      .eslintcache
      .eslintignore
      .eslintrc`;
    const current = `
      *.pid.lock
      *.seed
      *.seed
      *.tgz
      *.log
      .eslintignore
      .eslintrc
      media/
      node_modules/
      src/`;

    expect(
      mergeFiles(
        {
          path: '',
          hash: '',
          executable: false,
          type: FileType.GLOB,
          merge: true,
          content: source,
        },
        current
      )
    ).toMatchSnapshot();
  });
});
