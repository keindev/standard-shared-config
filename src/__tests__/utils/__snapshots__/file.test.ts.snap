// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File utils Get file hash 1`] = `"27a40f27cd71808962e218cfa221d14ff53b7f0d9db5ea56666d3422bb5d4e94"`;
